### ISSUE TITLE
Populate slow

### ISSUE THREAD

**Issue by RWOverdijk** [balderdashy/sails-mysql#161]
I have a model, with 3 associations. Fetch 30 records, with 3 associated records each causes sails to respond rather slow. (520ms with populate, 20ms with populate=false). The query itself takes no time at all so composing the objects is what's causing the slowness. I've tried optimising this by going trough the code and gave up because the code was a bit hard to follow.

**Comment by nbiton** [comment]
I have a feeling that this issue isn't directly related to the sails-mysql adapter.
I encountered similar results and noticed it is much more noticeable when the query is populating  associations of models that also have associations of their own.
I profiled and debugged waterline to see where this slowness comes from and managed to make some minor changes which made the processing much more efficient, so it made the processing time drop below 30% than what it was before, at least for my tests.

This is the relevant PR:
https://github.com/balderdashy/waterline/pull/829

@RWOverdijk It would be interesting to know if this has any impact on the results of your tests.

**Comment by devinivy** [comment]
+1 to that.  It would be nice to isolate various causes of this issue.

**Comment by lghamie** [comment]
+1 facing same issue

**Comment by devinivy** [comment]
This issue is further along over at https://github.com/balderdashy/sails-postgresql/issues/141

**Comment by dmarcelino** [comment]
Hey guys, I've tested the PR with the load test in [waterline-adapter-tests#load_test](https://github.com/balderdashy/waterline-adapter-tests/tree/load_test) agains sails-memory and the results were as follows:

branch | max memory consumed | time taken
----------|----------------------------------|---------------
master | 175.24 MB | 6776ms
nbiton  | 143.78 MB | 4478ms

I ran the tests 3 times in my machines and the results were consistent.

The change proposed by @nbiton speeds up `populate()` and saves some precious memory. I've also ran the integration tests and the change doesn't introduce any issue. We should look into merging this.

**Comment by sailsbot** [comment]
Thanks for posting, @RWOverdijk.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by tarlepp** [comment]
Hmm, is this issue solved?

I think that bot is too aggressive to close "old" issues, specially when users cannot re-open these issues.

**Comment by devinivy** [comment]
Reopened!  If anyone has new results or suggestions, please feel free to share here.

**Comment by RWOverdijk** [comment]
There was word of a faster populate, but I still think it could be made faster :) If this is a worry to more people. I think we should tackle it. The issue now is with how strings are used as identifiers, and new objects get constructed for all entities (logical)

**Comment by nbiton** [comment]
I have made an additional attempt to make populate queries faster a few months back.
Benchmarks I made on Node v0.10.x have shown a significant improvement, but results with Node v0.12.x were not as good for some reason.

I feel that I need to investigate further and see how my code fares in other scenarios, so I don't think it's ready yet for a PR, but it didn't break any tests.
The only functional difference I could see this has over the existing code is that populated records which are later saved with the "save"  method will always attempt to update the populated sub-records, regardless of whether they were changed or not.

This is the fork I created:
https://github.com/nbiton/waterline/tree/query-optimize

**Comment by sailsbot** [comment]
Thanks for posting, @RWOverdijk.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by tarlepp** [comment]
Could some one turn of this annoying bot, who is closing issues that aren't clearly solved. OR even better could core devs do something to these issues?

**Comment by devinivy** [comment]
I feel your pain.  CC @sailsbot @mikermcneil @particlebanana @tjwebb

**Comment by RWOverdijk** [comment]
:+1:

**Comment by mikermcneil** [comment]
@tarlepp @devinivy working on making sailsbot learn about labels

**Comment by Zaggen** [comment]
@mikermcneil If it isn't resolved and @sailsbot is still working on learning good manners, then shouldn't this issue be re-opened?

**Comment by particlebanana** [comment]
@zaggen I can re-open this but it's still not clear this is a bug in sails-mysql specifically. It's a very generic conversation with many side conversations, some of which have been resolved.

**Comment by Zaggen** [comment]
@particlebanana thanks, and I get what you mean, but since it is unclear where is the actual issue, it's probably better to have it open until we can be sure is solved or that is not directly related to sails-mysql, or at least that's what i think.

**Comment by RWOverdijk** [comment]
I just checked this again. 14ms when populate is null vs 290ms when populating three (small) models. This is in the browser, localhost, but should give you an idea about how much slower it is (~21 times, or ~2100% slower).

**Comment by sailsbot** [comment]
Thanks for posting, @RWOverdijk.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by tarlepp** [comment]
And again this issue is not solved, but @sailsbot closes this anyway... Could someone disable that annoying bot, it's not doing any good...

**Comment by jianpingw** [comment]
I feel that this issue is one of the most serious issues of Waterline. I have a very small data set (~100 records) and populating 3 associated records takes about 3 seconds with huge memory footprint.

I use Postgresql and I can confirm that the native query done by Postgresql adapter is extremely fast. The performance and memory issue seem to come from converting the rows into JSON format later on. It looks to me the algorithm is O(N*N) in terms of speed and memory usage.

Waterline has very nice APIs, but unfortunately the populating implementation severely affecting its usefulness.

**Comment by RWOverdijk** [comment]
Hazaa. I wish I could tackle this right now, but I'm a bit swamped. I know it has something to do with loops, and populating and... More. If anyone else feels like a hero, I'd love to work together on this. Otherwise unfortunately we'll have to wait until I have more time for this.

**Comment by jianpingw** [comment]
I did some instrumentation and for my case, this line takes about 50% of the total time (the total time is about 10 seconds):
https://github.com/balderdashy/waterline/blob/master/lib/waterline/query/finders/basic.js#L373

And these lines take about the other 50%:
https://github.com/balderdashy/waterline/blob/master/lib/waterline/query/finders/basic.js#L387-L389

**Comment by devinivy** [comment]
@jianpingw thanks for looking this over!  Very useful!!

**Comment by jianpingw** [comment]
@devinivy I will strip down our code and provide a test case, maybe in couple days.

**Comment by RWOverdijk** [comment]
@jianpingw +:100: very useful information

**Comment by jianpingw** [comment]
Here is the test case:
https://github.com/jianpingw/wl-perf

In my actual application, it takes about 10 seconds. However it only takes about 3 seconds in the test case.

**Comment by particlebanana** [comment]
From a quick glance looks like a (few) slow synchronous loops and slowness instrumenting the model. Obviously the more models and data you add the slower it will become because it has to take each record and create a new model for it. It would be great to speed all that up!

**Comment by devinivy** [comment]
Creating a single model instance is currently expensive.  It does a bunch of calculations that could be pre-computed and live on the collection.  I'll poke around more and look for low-hanging fruit.

**Comment by jianpingw** [comment]
Looks like my feeling of O(N*N) is correct. For this test case:
https://github.com/jianpingw/wl-perf

It finds 100 "message" models, each "message" model has five associations. So it should have 5*100=500 models.
 
However, the following lines are producing 5*100*100 (50000) models:
https://github.com/balderdashy/waterline/blob/master/lib/waterline/query/finders/joins.js#L167-#L195

Simply add a line at line 170:
`console.log('key: ' + key + '  value[key].length: ' + value[key].length);`

And run the test case. In this case, all the 100 items in `value[key]` are identical. If we can figure out why the algorithm is O(N*N) instead of O(N), we can significantly improve the speed.

**Comment by jianpingw** [comment]
I submitted two pull requests, one for MySQL and the other for Postgresql. I have seen 15x performance improvement with this fix.

I am new to WL, so this could be a naive fix.

**Comment by particlebanana** [comment]
@jianpingw thanks for the detailed dive! I'll take a look at the PR's.

**Comment by Zaggen** [comment]
Awesome guys, specially @jianpingw +:100: , @particlebanana, i hope we get good news after the inspection :)

**Comment by particlebanana** [comment]
Alright everyone this has been published and I think it officially closes out this issue based on the title **Populate slow**. A 15x performance boost solves that I think. Enjoy your weekend!

**Comment by tailot** [comment]
Very Nice!!! Thankssss

**Comment by devinivy** [comment]
This rocks– thanks again to @particlebanana and @jianpingw for getting that through and published.

I want to add that it's still very much worth looking into further perf improvements– in the adapters, waterline-cursor, and waterline core especially.  This issue does have some good ideas about where to look.

**Comment by tarlepp** [comment]
Simple table of improvements would be very nice. So basically columns: query, old, new, pure. I'm very happy to hear that there is some improvements for speed but I still think that pure solutions are way more faster.

So please provide us this simple perf table to show what is the speed improvement.

**Comment by Zaggen** [comment]
Thanks again to @jianpingw, @particlebanana and everybody else. 10-15x improvement is a lot, but as @devinivy and @tarlepp suggest, there might still be room for improvement. Maybe opening a new issue and tagging it as a "feature" instead of like a "bug" and linking it to this issue, as opposed to reopening this one could be a nice idea, or what do you guys think?, In any case, i'm thankful for this improvement(I haven't tested it though) :+1:
